### PR TITLE
Adding a cacheWithRedis() wrapper + make memoize LRU based

### PR DESCRIPTION
**Pull request by lasryaric**
# Description

Moving away from using the memoize() function in connectors because it's just not viable. We are now using the Redis cache instead.

The cacheWithRedis() function works like memoize, but has an explicit name.

The memoize() function is now LRU based (vs time based) and takes a `maxCapacity` argument.
The memoize() function is not used anywhere today but since it was a lot of work to get the typing properly working and it is a common function, I would vote for keeping it. If you think we should remove it, LMK.

# Deployment
- Create the new redis cache server
- Set the var env `REDIS_CACHE_URI`
- deploy connectors.

### CODE REVIEW

File: types/package.json [hunk 1]
```diff
@@ -30,6 +30,11 @@
     "dts-cli": "^2.0.3",
     "eventsource-parser": "^1.1.1",
     "io-ts": "^2.2.20",
+    "redis": "^4.6.8",
     "uuid": "^9.0.1"
+  },
+  "browser": {
```
Comment (spolu) [inline types/package.json:36]: please give us a crash course on this in the PR comments :)

File: types/src/shared/memoize.ts [hunk 2]
```diff
@@ -1,29 +1,21 @@
 // Memoize function with TTL.
 // Usage:
-// const memoizedFn = memoize(fn, (fnArg1, fnArg2, ...) => `${fnArg1}-${fnArg2}`, 60 * 10 * 1000);
-// Cache expires after ttlMs since last usage of the memoized function.
+// const memoizedFn = memoize(fn, (fnArg1, fnArg2, ...) => `${fnArg1}-${fnArg2}`, 10);
+// The keys are evicted in LRU order after the cache reaches maxCapacity.
 export function memoize<T extends (...args: any[]) => any>(
```
Comment (spolu) [inline types/src/shared/memoize.ts:5]: can we remove that file?

File: types/src/shared/cache.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      throw new Error("REDIS_CACHE_URL is not set");
+    }
+    redisUri = REDIS_CACHE_URI;
+  }
+
+  return async function (...args: Parameters<T>) {
+    if (!redisUri) {
+      throw new Error("redisUrl is not set");
+    }
+    const a: ReturnType<T> | undefined = undefined;
```
Comment (spolu) [inline types/src/shared/cache.ts:24]: warning `a` never used

File: types/src/shared/cache.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  if (!redisUri) {
+    const REDIS_CACHE_URI = process.env.REDIS_CACHE_URI;
+    if (!REDIS_CACHE_URI) {
+      throw new Error("REDIS_CACHE_URL is not set");
+    }
+    redisUri = REDIS_CACHE_URI;
+  }
+
+  return async function (...args: Parameters<T>) {
+    if (!redisUri) {
```
Comment (spolu) [inline types/src/shared/cache.ts:21]: is this possible as per above?

File: types/src/shared/cache.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import { redisClient } from "../shared/redis_client";
+
+// Wrapper function to cache the result of a function with Redis.
+// Usage:
+// const cachedFn = cacheWithRedis(fn, (fnArg1, fnArg2, ...) => `${fnArg1}-${fnArg2}`, 60 * 10 * 1000);
+export function cacheWithRedis<T extends (...args: any[]) => Promise<any>>(
+  fn: T,
+  resolver: (...args: Parameters<T>) => string,
+  ttlMs: number,
```
Comment (spolu) [inline types/src/shared/cache.ts:9]: Let's refuse more than 1 day by throwing an error for defense in depth

**Review by spolu — COMMENTED** [review]
LGTM \o/

But I want to understand `browser` stuff before merging :pray:

File: types/src/shared/cache.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import { redisClient } from "../shared/redis_client";
+
+// Wrapper function to cache the result of a function with Redis.
+// Usage:
+// const cachedFn = cacheWithRedis(fn, (fnArg1, fnArg2, ...) => `${fnArg1}-${fnArg2}`, 60 * 10 * 1000);
+export function cacheWithRedis<T extends (...args: any[]) => Promise<any>>(
```
Comment (fontanierh) [inline types/src/shared/cache.ts:6]: does it work if we replace the 2 `any` by `unknown` ?

File: types/src/shared/cache.ts (hunk 4 above)
Comment (fontanierh) [reply types/src/shared/cache.ts:21]: (Def ok with the second check)

File: types/src/shared/memoize.ts [hunk 7]
```diff
@@ -1,29 +1,21 @@
...
-    clearTimeout(ttlTimeout);
-    ttlTimeout = setTimeout(() => {
-      cache.clear();
-    }, ttlMs);
-  };
-  extendTTL();
   return function (...args: Parameters<T>): ReturnType<T> {
-    extendTTL();
-
     const key = resolver(...args);
```
Comment (fontanierh) [inline types/src/shared/memoize.ts:13]: If I understand correctly, different memoize fns can collide / use the same keys, right ?

**Review by fontanierh — DISMISSED** [review]
This is looking great. I have some questions, but nothing blocking

File: types/package.json (hunk 1 above)
Comment (lasryaric) [reply types/package.json:36]: My understanding is that there is a build step in nextjs that was trying to package the file importing Redis in a front-end code, which is not possible, and adding that tells the builder (webpack apparently) to not package these for client side code.

Unfortunately, online documentation is really poor on this, so I can't provide a super clear answer.  But everything works well.

This is the most plausible answer I found [online](https://stackoverflow.com/a/74240581):

When I added redis as a dependency to `types/`, front build started to fail with the following errors:
```
Module not found: tls
Module not found: net
```
The solution to this error was to specify that.

**Review by lasryaric — COMMENTED** [review]
addressed all your comments.

File: types/package.json (hunk 1 above)
Comment (spolu) [reply types/package.json:36]: OK let's give it a try. I worry we might be missing with SSR here... but we'll know soon enough. Let's just remind ourselves to revisit this if we feel there are performance regressions in the coming days

**Review by spolu — APPROVED** [review]

File: types/src/shared/cache.ts (hunk 6 above)
Comment (lasryaric) [reply types/src/shared/cache.ts:6]: it won't. the caller sees a function with the same type as the one passed as a parameter. It's just that this function can take any type of function, so it has to be any.

**Pull request merged**